### PR TITLE
Add integration tests for sql cmd

### PR DESCRIPTION
**Pull request by carlosms**
Part of #232.
Adds tests for `srcd sql`.

### CODE REVIEW

**Comment by se7entyse7en** [comment]
CI failing.

File: cmd/srcd/cmd/sql_test.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	command.Stderr = &out
+	command.Stdin = &in
+
+	io.WriteString(&in, "show tables;\n")
+	io.WriteString(&in, "describe table repositories;\n")
+
+	err := command.Start()
+	require.NoError(err)
+
+	err = command.Wait()
```
Comment (smacker) [inline cmd/srcd/cmd/sql_test.go]: could you please help me to understand how it works.
`.Wait` should block until command exits. But `repl` doesn't exit by itself and I don't see where you send stop signal to it.

File: cmd/srcd/cmd/sql_test.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+| repositories |
+| tree_entries |
++--------------+
++---------------+------+
+|     NAME      | TYPE |
++---------------+------+
+| repository_id | TEXT |
++---------------+------+`
+
+	require.Contains(out.String(), expected)
```
Comment (smacker) [inline cmd/srcd/cmd/sql_test.go]: I would also add test writing to stdin after we got results from the previous command. Just to simulate more real use of the repl.

File: cmd/srcd/cmd/sql_test.go (hunk 1 above)
Comment (carlosms) [reply cmd/srcd/cmd/sql_test.go]: Yes, this is tricky, I will add some comments to the code.

What I _think_ is happening is that the [readline lib](github.com/chzyer/readline) we use must detect that it is not in a terminal, and exits when sdtin is consumed.

You can see it if you run the bin like this. It runs the query, prints the result and exits.
```
$ echo "show tables;" | ./srcd sql
```

So we are still testing [the code that processes](https://github.com/src-d/engine/blob/master/cmd/srcd/cmd/sql.go#L108) the lines read by the repl, but without interaction. There are things we can't test, like exit on `exit;`, but at least this will fail if we break something major there.

File: cmd/srcd/cmd/sql_test.go (hunk 2 above)
Comment (carlosms) [reply cmd/srcd/cmd/sql_test.go]: Related to the other conversation: https://github.com/src-d/engine/pull/314#discussion_r264658396

Because the command reads stdin and exits, we are not really in interactive mode, and I think we can't test what you describe.

File: cmd/srcd/cmd/sql_test.go (hunk 1 above)
Comment (carlosms) [reply cmd/srcd/cmd/sql_test.go]: Added comments in 1a374ba, will squash it before merging

**Comment by smacker** [comment]
Thanks for the explanation!

Most probably we still can test `exit` and other stuff if we run the command as
```
/usr/bin/bash -e "/path/to/srcd sql"
```

but better to do it in a separate PR.

File: cmd/srcd/cmd/sql_test.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+	buf, err := s.RunSQL(context.TODO(), "show tables")
+	require.NoError(err)
+
+	require.Contains(buf.String(), showTablesOutput)
+}
+
+func (s *SQLTestSuite) TestValidQueries() {
+	testCases := []string{
+		"show tables",
```
Comment (dpordomingo) [inline cmd/srcd/cmd/sql_test.go]: isn't this guaranteed by gitbase itself? If so, why replicating tests here?

File: cmd/srcd/cmd/sql_test.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+func (s *SQLTestSuite) TestWrongQuery() {
+	testCases := []struct {
+		query string
+		err   string
+	}{
+		{
+			query: "show",
+			err:   "rpc error: code = Unknown desc = SQL query failed: Error 1105: unknown error: syntax error at position",
```
Comment (dpordomingo) [inline cmd/srcd/cmd/sql_test.go]: do we want to ensure that error is **exactly** that? even with error number and formating?

File: cmd/srcd/cmd/sql_test.go (hunk 3 above)
Comment (carlosms) [reply cmd/srcd/cmd/sql_test.go]: gitbase should work for `show tables`, but we need to test that `srcd` is in fact sending the query as expected.
I put some variations to check that we don't mess with the query string. This could happen if for example tomorrow we decide to trim spaces and add a `;` to all queries, this test would detect some corner cases.

File: cmd/srcd/cmd/sql_test.go (hunk 4 above)
Comment (carlosms) [reply cmd/srcd/cmd/sql_test.go]: what do you propose?
My thinking here is that we want to check that the error is in fact a syntax error, not some other random crash. For example if we just checked for an error, it could happen that our client crashes with a nil pointer panic when gitbase returns an error; and the test would pass.

File: cmd/srcd/cmd/sql_test.go (hunk 3 above)
Comment (dpordomingo) [reply cmd/srcd/cmd/sql_test.go]: should we rewrite the query sent? should not `gitbase` to receive it as it is? why using source{d} Engine to polish sql queries?
I agree that we need to test that `srcd` sends the query, and gets the proper result, but imo it is done just sending a query, and comparing with the expected result; caring about casing, spaces or other formatting things, should be done by `gitbase`.

Anyhow, it's not a blocker to merge this PR, just a personal opinion ;)

File: cmd/srcd/cmd/sql_test.go (hunk 4 above)
Comment (dpordomingo) [reply cmd/srcd/cmd/sql_test.go]: I agree that we must test that it is caused a syntax error, and not a random one; what I would suggest, is not caring about the error format. Maybe just searching for `syntax error at position` would be enough because the rest is added by different layers of Engine (which do not add extra value to the user), so imo it should not be validated in tests.

Comment (carlosms) [reply cmd/srcd/cmd/sql_test.go]: I updated the test to check only the inner error, ignoring the `rcp error` prefix. Since it's a small change, I already squashed it.
